README.adoc
= NMEA2000 library for C++ =

image:https://badges.gitter.im/docu-for-NMEA2000-C-lib/community.svg[link="https://gitter.im/docu-for-NMEA2000-C-lib/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge"]

Object oriented NMEA2000 library for Teensy, ESP, Arduino, MBED and Rasberry type Boards.
These board types has been tested, but library can be used also in other systems by writing
compatible CAN driver and wrapper for other hw specific functions.

Library gives you easy way to make different kind of NMEA2000 bus devices like
sensor transducers (battery, temperature, wind, engine, etc.), NMEA2000 information displays,
NMEA2000->PC interface (like Actisense NGT1), NMEA0183->NMEA2000 or NMEA2000->NMEA0183 converter.

Library fulfills NMEA 2000 mandatory functions and behaviour. Devices using library can pass NMEA2000
certification tests. Library has been used in several commercial certified products.

== Documents ==

Please read document https://github.com/ttlappalainen/NMEA2000/blob/master/Documents/Connecting_hardware_to_NMEA2000.pdf
before purchasing any hardware.

For library there is document https://github.com/ttlappalainen/NMEA2000/blob/master/Documents/NMEA2000_library_reference.pdf

== Recommended hardware ==

I prefer to use *Teensy 3.2, 3.5, 3.6, 4.0, 4.1* or *ESP32* board with either unisolated 
*MCP2562* or isolated *ISO1050* tranceiver. Those boards are "very" powerfull, small and
runs with low power. See more about hardware under <<Hardware setup>> below.

With default settings library requires about *23 kB rom and 3.3 kB RAM* in normal operation. So you should
have at least *8 kB RAM* in your chip. If you have Arduino Mega board, it is OK for testing and for small projects,
but I do not prefer to buy one for new project. Arduino Due is OK, but is physically bigger and eats
more power than Teensies or ESP32. If you have Arduino Uno or any board with less than 8 kB RAM, please read
all issues and <<Memory requirements>> below and then forget it. If you anyway want to use it
please do not set any issues, if it does not work or does unexpected crashes.

== Hardware depended libraries ==

Take care that you use libraries under my Github, when available! Others may not work right
with NMEA2000!

=== Teensy 4.0/4.1 boards with internal CAN ===
 
- https://github.com/ttlappalainen/NMEA2000_Teensyx[NMEA2000_Teensyx] library.

CAN library is included to the code so you do not need any extra CAN library with this. NMEA2000_Teensyx
library will replace NMEA2000_teensy library in future. You can already start to use it with
all Teensy boards by forcing it with define (see NMEA2000_CAN.h comment). For critical
projects I prefer to use old library until this has been running under tests for a while.
 
=== Teensy 3.1/3.2 or 3.5/3.6 board with internal CAN ===

- https://github.com/ttlappalainen/NMEA2000_teensy[NMEA2000_teensy] library.

- https://github.com/ttlappalainen/FlexCAN_Library[FlexCAN] library.

NMEA2000_teensy has been originally forked from https://github.com/sarfata/NMEA2000_teensy[NMEA2000_teensy] contributed
by (https://twitter.com/sarfata/[Thomas Sarlandie])

*Note* that FlexCan library is already included in
https://www.pjrc.com/teensy/teensyduino.html[Teensyduino] extension. In order
to use my fork, you have to remove FlexCAN library from `C:\Program Files
(x86)\Arduino\hardware\teensy\avr\libraries\FlexCAN`. Default FlexCAN library does not
work right with NMEA2000.

=== ESP32 boards  with internal CAN ===

- https://github.com/ttlappalainen/NMEA2000_esp32[NMEA2000_esp32] library.

ESP32 CAN driver by Thomas Barth has been implemented inside NMEA2000_esp32
so you do not need other drivers for that.

=== Arduino Due board with internal CAN ===

- https://github.com/ttlappalainen/NMEA2000_due[NMEA2000_due] library

- https://github.com/ttlappalainen/due_can[due_can] library

=== Boards with MCP2515 CAN bus controller (e.g. Arduino Mega, ESP8266) ===

- https://github.com/ttlappalainen/NMEA2000_mcp[NMEA2000_mcp] library.

- https://github.com/ttlappalainen/CAN_BUS_Shield[mcp_can] library.

You can use mcp_can with boards where you connect MCP2515 CAN bus controller with SPI. You can
also use it as secondary CAN for e.g. Teensy 3.2 or ESP32.

mcp_can library is developed version of original mcp_can library. 
This allows also to use 8MHz clock and has been tested with
Maple Mini.  mcp_can library can be originally found on
https://github.com/Seeed-Studio/CAN_BUS_Shield, but check library status has all improvements
been implemented to Seeed-Studio. At least 18.09.2017 it was up to date. 

=== avr processors with internal CAN ===

- https://github.com/thomasonw/NMEA2000_avr[NMEA2000_avr] library

- https://github.com/thomasonw/avr_can[avr_can] library

=== MBED boards ===

- https://github.com/thomasonw/NMEA2000_mbed[NMEA2000_mbed] library

- other related libraries. See origin for MBED port on https://github.com/thomasonw/NMEA2000_mbed

=== Raspberry Pi ===
 
- https://github.com/ttlappalainen/NMEA2000_socketCAN[NMEA2000_socketCAN] library.

NMEA2000_socketCAN has been originally forked from https://github.com/thomasonw/NMEA2000_socketCAN

== Thanks ==

I thank Kees Verruijt for his excellent CanBoat project and Dr Andrew Mason for
OpenSkipper project.  I originally started to learn NMEA 2000 via OpenSkipper
project, which development I continued.  PGN library on OpenSkipper and so on
this library is mostly based on Kees work. Also it appeared that some code
parts are probably originally from CanBoat even I have found them from other
sources.  Actisense message format handling I have learned from OpenSkipper
project.

I also thank for anybody who has extended library with new PGNs or processor
support.

== Changes ==
11.08.2022

- Fix: On N2kGroupFunctionDefaultHandlers.cpp tN2kGroupFunctionHandlerForPGN60928 unique number length is 3 bytes.

- Fix: On N2kMsg.cpp changed round(v/precision) to round((v/precision)), since some system round macro behaved unexpectedly. Thaks to ieb.

06.07.2022

- Fix: use proper array delete operator in RingBuffer, thanks garaminowicz

- Fix: PGN 127251 rate of turn type, thanks BerndCirotzki and for remind speters.

- Fix: PGN 129794 setter.

== Changes ==
07.04.2022

- Changed #include <cstring> to #include <string.h>. <cstring> is not available on all environments.

- Fixed proprietary PGN handling on N2kGroupFunction.cpp.

- Fixed N2kMsg.cpp:SetBuf3ByteDouble(). Thanks to mairas.

- Add support for PGN 129041, 130323. Thanks to MattCairns.

- Added support for PGN127237, 127233, 130577 and more detail for 129039. Thanks to aydosj.

- Copyright update.

05.09.2021

- Documentation update.

23.08.2021

- Added manufacturer bits to engine discrete status ( PGN 127489 )

- Added data limitation handling to tN2kMsg. Now overflow values will be sent as overflow. E.g.
  SetN2kTemperature(N2kMsg,0xff,0,N2kts_MainCabinTemperature,CToKelvin(800)); +
  will send temperature as 0xfffe, which is NMEA2000 standard way to inform that value is out of range.
  Library does not yet handle overflow on reading, but it is under construction.
  
- Fixed PGNs 126992, 128275, 129029 time reading to UDouble

- Added function tNMEA2000::Restart(). This is preliminary and may change. Read more info on NMEA2000.h.

- Modified tNMEA2000::RespondGroupFunction to allow user add own common group function handler with PGN=0.

- Fixed tNMEA2000::HandleCommandedAddress which should be accepted only by BAM TP.

- Added tN2kGroupFunctionHandler::ChangeTransmissionOrPriorityErrorCode

- Fixed tN2kGroupFunctionHandlerForPGN126993::HandleRequest. System must not respond to pure 126208 request for 126993.
  Also TransmissionInterval setting is limited to 1000 - 60000 ms.
  
- Modified debug messages.

04.03.2021

- Fix for engine discrete status 2.

10.02.2021

- Added definition to NMEA2000_CAN.h to select CAN device for Arduino DUE

06.02.2021

- Added handlers for PGN 129540 GNSS satellites in view

- Added new necessary types

- Added SetByte for tN2kMsg class.

- Added PGN 129540 handler for test example DataDisplay2

- Redesigned example MessageSender. Now it is more usefull for testing, since each
  message can be individually enabled/disabled.

- Added PGN 129540 handler for test example MessageSender

15.01.2021

- Added better inline ParseN2kEngineDynamicParam to N2kMessages

- Copyright update

05.01.2021

- Added better engine discrete status handling to N2kMessages

08.10.2020

- Fix for 'memcpy' overflow for float reading on N2kMsg.cpp

23.08.2020

- Add support for Teensy 4.x

05.08.2020

- Add support for Windlass Network Messages PGN 128776, 128777 & 128778 by Paul Reeve

16.07.2020

- Fixed setting buffered frame length to min instead of max length on NMEA2000.cpp SendFrame

- Added NMEA2000::IsProprietaryMessage

- Fixed AddGroupFunctionHandler to allow to add handlers after Open

- Added RemoveGroupFunctionHandler to allow dynamically remove handlers on runtime.

- Added proprietary message handling to N2kGroupFunction.cpp

- Improved N2kDeviceList HandleMsg to see devices, which has been off.

- Added FindDeviceByProduct to N2kDeviceList

- Extented list of fast packet messages to all known fast packet messages. This also leads to situation that all
  fast packet messages are known messages and may effect to setting SetHandleOnlyKnownMessages behaviour. To limit
  known messages, one should have provided list with SetFastPacketMessages.

27.06.2020

- Fixed PGN 130314 and PGN 130315 pressure type to signed.

- Added pressure enums.

28.01.2019

- NOTE! Compatibility change! PGN 127506 TimeRemaining should have been in seconds as it is SI unit.
  If you have used that on your code, provide value in seconds. Parse function also return in seconds now.
  
- Added Capacity parameter to PGN 127506. I did not made overwrite function without due to above change. 
  So it hopefully wakes you up about the changes.
  
- Fixed MaretronProprietary test on N2kMaretron.cpp

- Some comment fixes.

20.10.2019

- Added support for Maretron proprietary PGNs 130823,65286, 65287. See N2kMaretron.h.
  Thanks to Vassilis Bourdakis.
  
- Added proprietary fast packet message test as default. So now proprietary fast packet
  messages will be automatically parsed right without need to use ExtendFastPacketMessages
  
- Fixed strings on product information. Unused characters will be filled with 0xff.

17.10.2019

- Fixed Rate of turn (PGN 127251) and AIS position report ROT (PGN 129038) value multiplier.

- Separated N2k enums to own file N2kTypes.h

- Fixed PGNs 127251, 127258, 130576 length for sending by padding with reserved. Some devices refuces
  to listen message, if length is wrong.

07.07.2019

- Fixed: PGN 127513 was accidently defined also as single frame message.

07.07.2019

- Added: Trip fuel consumption, engine PGN 127497.

- Added: More PGN:s to default fast packet list.

- Updated: Examples MessageSender and DataDisplay2, which can be used for testing messages.

03.03.2019

- Added: Charger status PGN 127507.

- Added: Possibility to delay ISO address claim. Due to some devices CAN priority, it was sent too fast.

- Added: Count for tDeviceList

14.01.2019

- Fix: Device list handler fix. Some tools may use source 254 and that was checked wrong.

- Fix: Message priorities.

04.08.2018

- Fix: NMEA2000.h/NMEA2000.cpp, just in parameter N2kSource type change by mrbubble62

18.07.2018

- Added: N2kMessagesEnumToStr.h "exhaust gas" string for temperature sources by mrbubble62.

09.05.2018

- Fix: NMEA2000_CAN.h, make ESP32 work with ESP-IDF framework, by Sarfata

10.04.2018

- Fix: Handling of humidity PGN 130313.

06.04.2018

- Added: Support for ESP32. See also NMEA2000_esp32

- Fix: Changed some names to avoid conflicts with some stupid define macros on some environments.

04.04.2018

- Fix: NextHeartbeatSentTime initial value.

- Added: New example NMEA2000ToNMEA0183. This has been tested on RPi3B, Arduino DUE, Arduino Mega, Teensy.

26.03.2018

- Added: Strings for Magnetic Variation enum by mrbubble62

13.03.2018

- Added: PGN129033 Local offset

23.02.2018

- Port config ability for SockeCAN by Al Thomason

16.02.2018

- SetN2kSource for other devices by jpilet

29.01.2018

- Added: PGN130576 Small Craft Status / Trim Tab Position definition by Nicholas Agro

15.01.2018

- Fix: ParseN2kPGN129284, Index was not initialized to 0, which caused unpredictable read.

07.01.2018

- Fix: Fast packet sequence counter must be related to PGN.

- Fix: TP message priorities.

- Fix: On SetDeviceInformationInstances we need to send ISO address claim, not start adress claim.

- Fix: Do not respond any queries during address claim.

- Fix: Fixed some messages default priority.

- Fix: Responce to Complex Group Function requests.

- Fix: Now compiles with different compiler definitions defined on NMEA2000_CompilerDefns.h

- Added: Support for sending messages by using Transport Protocol. This is mandatory for NMEA 2000

- Updated: License to 2018

02.12.2017

- Changes effect only use of tN2kDeviceList and AttachMsgHandler callbacks.

- Added FindDeviceByIDs to tN2kDeviceList

- Fix: Devicelist did not handle right, if device changed its address higher.

- Fix: PNG message handler order problem on tNMEA2000::AttachMsgHandler(tMsgHandler *_MsgHandler);

- Fix: If there was NMEA2000 library device with same name, they handshaked both addresses to null.
  The problem still appears, if both devices will be started at same time. I need to add some
  random start delay
  
- Cleaned spaces from end of lines on updated code.

29.11.2017

- Added PGN 128000 Leeway to N2kMessages.h/.cpp

- Fix: Spelling WaterRefereced -> WaterReferenced on N2kMessages.h/.cpp

27.11.2017

- Fix some build errors due to missing include statements (did not affect Arduino, only other platforms)

14.11.2017

- Fix: Default responces to Complex Group Function.

09.11.2017

- Added library.json, thanks for ronzeiller.

31.10.2017

- Fix: fill unused chars on fastpacket messages with 0xff

- Fix: fill unused chars on message fixed sized strings with 0xff. One sample is e.g. PGN 126996, Product Information.

14.10.2017

- Changed tActisenseReader class to handle also Actisense N2k request message types. This type will be used
  by applications, which sends data through NGT-1.
  
- Updated related examples ActisenseListener and ActisenseListenerSender.

- Renamed Min/Max to N2kMin/N2kMax, since Due code had definitions for Min/Max

12.10.2017

- Added PGN validity check to the SendMsg()

- Changed max/min -> own Max/Min. max/min defines does not exist on other systems and Arduino does not have std <Algorithm> for all boards.

10.10.2017

- Fix: PGN 126992 TimeSource handling.

- Fix: Added missing wind reference type.

02.10.2017

- Added range parameter for depth PGN 128267. 

- Added millis() time stamp to N2kMsg clear text Print.

- Now responces also Group Function request for PGN lists PGN 126464

- Now responces also Group Function request for Product Information PGN 126996


docu/changes.md
\page changes Changes to the Library

11.08.2022

- Fix: On N2kGroupFunctionDefaultHandlers.cpp tN2kGroupFunctionHandlerForPGN60928 unique number length is 3 bytes.

- Fix: On N2kMsg.cpp changed round(v/precision) to round((v/precision)), since some system round macro behaved unexpectedly. Thaks to ieb.

06.07.2022

- Fix: use proper array delete operator in RingBuffer, thanks garaminowicz

- Fix: PGN 127251 rate of turn type, thanks BerndCirotzki and for remind speters.

- Fix: PGN 129794 setter.


07.04.2022

- Changed #include <cstring> to #include <string.h>. <cstring> is not available on all environments.

- Fixed proprietary PGN handling on N2kGroupFunction.cpp.

- Fixed N2kMsg.cpp:SetBuf3ByteDouble(). Thanks to mairas.

- Add support for PGN 129041, 130323. Thanks to MattCairns.

- Added support for PGN127237, 127233, 130577 and more detail for 129039. Thanks to aydosj.

- Copyright update.

05.09.2021

- Documentation update.

23.08.2021

- Added manufacturer bits to engine discrete status ( PGN 127489 )

- Added data limitation handling to tN2kMsg. Now overflow values will be sent as overflow. E.g.
  SetN2kTemperature(N2kMsg,0xff,0,N2kts_MainCabinTemperature,CToKelvin(800)); +
  will send temperature as 0xfffe, which is NMEA2000 standard way to inform that value is out of range.
  Library does not yet handle overflow on reading, but it is under construction.
  
- Fixed PGNs 126992, 128275, 129029 time reading to UDouble

- Added function tNMEA2000::Restart(). This is preliminary and may change. Read more info on NMEA2000.h.

- Modified tNMEA2000::RespondGroupFunction to allow user add own common group function handler with PGN=0.

- Fixed tNMEA2000::HandleCommandedAddress which should be accepted only by BAM TP.

- Added tN2kGroupFunctionHandler::ChangeTransmissionOrPriorityErrorCode

- Fixed tN2kGroupFunctionHandlerForPGN126993::HandleRequest. System must not respond to pure 126208 request for 126993.
  Also TransmissionInterval setting is limited to 1000 - 60000 ms.
  
- Modified debug messages.

04.03.2021

- Fix for engine discrete status 2.

10.02.2021

- Added definition to NMEA2000_CAN.h to select CAN device for Arduino DUE

06.02.2021

- Added handlers for PGN 129540 GNSS satellites in view

- Added new necessary types

- Added SetByte for tN2kMsg class.

- Added PGN 129540 handler for test example DataDisplay2

- Redesigned example MessageSender. Now it is more usefull for testing, since each
  message can be individually enabled/disabled.

- Added PGN 129540 handler for test example MessageSender

15.01.2021

- Added better inline ParseN2kEngineDynamicParam to N2kMessages

- Copyright update

05.01.2021

- Added better engine discrete status handling to N2kMessages

08.10.2020

- Fix for 'memcpy' overflow for float reading on N2kMsg.cpp

23.08.2020

- Add support for Teensy 4.x

05.08.2020

- Add support for Windlass Network Messages PGN 128776, 128777 & 128778 by Paul Reeve

16.07.2020

- Fixed setting buffered frame length to min instead of max length on NMEA2000.cpp SendFrame

- Added NMEA2000::IsProprietaryMessage

- Fixed AddGroupFunctionHandler to allow to add handlers after Open

- Added RemoveGroupFunctionHandler to allow dynamically remove handlers on runtime.

- Added proprietary message handling to N2kGroupFunction.cpp

- Improved N2kDeviceList HandleMsg to see devices, which has been off.

- Added FindDeviceByProduct to N2kDeviceList

- Extented list of fast packet messages to all known fast packet messages. This also leads to situation that all
  fast packet messages are known messages and may effect to setting SetHandleOnlyKnownMessages behaviour. To limit
  known messages, one should have provided list with SetFastPacketMessages.

27.06.2020

- Fixed PGN 130314 and PGN 130315 pressure type to signed.

- Added pressure enums.

28.01.2019

- NOTE! Compatibility change! PGN 127506 TimeRemaining should have been in seconds as it is SI unit.
  If you have used that on your code, provide value in seconds. Parse function also return in seconds now.
  
- Added Capacity parameter to PGN 127506. I did not made overwrite function without due to above change. 
  So it hopefully wakes you up about the changes.
  
- Fixed MaretronProprietary test on N2kMaretron.cpp

- Some comment fixes.

20.10.2019

- Added support for Maretron proprietary PGNs 130823,65286, 65287. See N2kMaretron.h.
  Thanks to Vassilis Bourdakis.
  
- Added proprietary fast packet message test as default. So now proprietary fast packet
  messages will be automatically parsed right without need to use ExtendFastPacketMessages
  
- Fixed strings on product information. Unused characters will be filled with 0xff.

17.10.2019

- Fixed Rate of turn (PGN 127251) and AIS position report ROT (PGN 129038) value multiplier.

- Separated N2k enums to own file N2kTypes.h

- Fixed PGNs 127251, 127258, 130576 length for sending by padding with reserved. Some devices refuces
  to listen message, if length is wrong.

07.07.2019

- Fixed: PGN 127513 was accidently defined also as single frame message.

07.07.2019

- Added: Trip fuel consumption, engine PGN 127497.

- Added: More PGN:s to default fast packet list.

- Updated: Examples MessageSender and DataDisplay2, which can be used for testing messages.

03.03.2019

- Added: Charger status PGN 127507.

- Added: Possibility to delay ISO address claim. Due to some devices CAN priority, it was sent too fast.

- Added: Count for tDeviceList

14.01.2019

- Fix: Device list handler fix. Some tools may use source 254 and that was checked wrong.

- Fix: Message priorities.

04.08.2018

- Fix: NMEA2000.h/NMEA2000.cpp, just in parameter N2kSource type change by mrbubble62

18.07.2018

- Added: N2kMessagesEnumToStr.h "exhaust gas" string for temperature sources by mrbubble62.

09.05.2018

- Fix: NMEA2000_CAN.h, make ESP32 work with ESP-IDF framework, by Sarfata

10.04.2018

- Fix: Handling of humidity PGN 130313.

06.04.2018


src/NMEA2000_CAN.h
/*
 * NMEA2000_CAN.h
 *
 * Copyright (c) 2015-2022 Timo Lappalainen, Kave Oy, www.kave.fi
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
*/

/*! \file   NMEA2000_CAN.h
 *  \brief  Automatic library selection according to the selected board

  Just include this <NMEA2000_CAN.h> to your project and it will
  automatically select suitable CAN library according to board
  selected on Arduino development environment. You can still force
  library by adding one of next defines before including NMEA2000_CAN.h:
 
  \code{cpp}
  #define USE_N2K_CAN 1  // for use with SPI and MCP2515 can bus controller
  #define USE_N2K_CAN 2  // for use with due based CAN
  #define USE_N2K_CAN 3  // for use with Teensy 3.1/3.2/3.5/3.6 boards
  #define USE_N2K_CAN 4  // for use with avr boards
  #define USE_N2K_CAN 5  // for use with socketCAN (linux, etc) systems
  #define USE_N2K_CAN 6  // for use with MBED (ARM) systems
  #define USE_N2K_CAN 7  // for use with ESP32
  #define USE_N2K_CAN 8  // for use with Teensy 3.1/3.2/3.5/3.6/4.0/4.1 boards
  \endcode

  <b>Depending of your board you will need to also install "driver" libraries:</b>  

  see also \ref secHWlib

  Arduino CAN shield (mcp_can) with MCP2515 chip:
    - https://github.com/ttlappalainen/CAN_BUS_Shield
    - https://github.com/ttlappalainen/NMEA2000_mcp
    
  Arduino DUE internal CAN:
    - https://github.com/ttlappalainen/due_can
    - https://github.com/ttlappalainen/NMEA2000_due
    
  Teensy 3.2-3.6 internal CAN:
    - https://github.com/ttlappalainen/FlexCAN_Library
    - https://github.com/ttlappalainen/NMEA2000_teensy
    
  Teensy 4.X internal CAN:
    - https://github.com/ttlappalainen/NMEA2000_Teensyx
    
  ESP32 internal CAN:
    - https://github.com/ttlappalainen/NMEA2000_esp32
  
  AVR:
    - https://github.com/thomasonw/avr_can
    - https://github.com/thomasonw/NMEA2000_avr
  
  MBED (note that there may be problem with fastpackets on MBED):
    - https://github.com/thomasonw/NMEA2000_mbed
    
  RPi socket CAN:
    -  https://github.com/thomasonw/NMEA2000_socketCAN
  

  <b>There are also library specific defines:</b>
  
  mcp_can:

    \code{cpp}
    #define N2k_SPI_CS_PIN 53  // Pin for SPI Can Select
    #define N2k_CAN_INT_PIN 21 // Use interrupt  and it is connected to pin 21
    #define USE_MCP_CAN_CLOCK_SET 8  // possible values 8 for 8Mhz and 16 for 16 Mhz clock
    \endcode

  esp32:

    \code{cpp}
    #define ESP32_CAN_TX_PIN GPIO_NUM_16
    #define ESP32_CAN_RX_PIN GPIO_NUM_4
    \endcode

  Teensyx:

    \code{cpp}
    #define NMEA2000_TEENSYX_CAN_BUS tNMEA2000_Teensyx::CAN2 // select CAN bus 2
    #define USE_NMEA2000_TEENSYX_FOR_TEENSY_3X // Force NMEA2000_TEENSYX also for Teensy 3.x boards
    \endcode

  Arduino DUE:
    
    \code{cpp}
    #define NMEA2000_ARDUINO_DUE_CAN_BUS tNMEA2000_due::CANDevice1  // Use CAN bus 1 instead of 0 for Arduino DUE
    \endcode

*/


#ifndef _NMEA2000_CAN_H_
#define _NMEA2000_CAN_H_

#include "N2kMsg.h"
#include "NMEA2000.h"


/**
 * \brief Use the mcp_can library
 * 
 * CAN_BUS_shield libraries will be originally found on 
 * <https://github.com/Seeed-Studio/CAN_BUS_Shield>
 * There is improved library, which branch can 
 * be found on https://github.com/peppeve/CAN_BUS_Shield.  
 * That works also with Maple mini and 8 MHz clock. Hopefully these
 * improvements will be applied to original library
 */
#define USE_N2K_MCP_CAN 1
/**
 * \brief Use the Arduino Due CAN Library
- [NMEA2000_due](https://github.com/ttlappalainen/NMEA2000_due) library
- [due_can](https://github.com/ttlappalainen/due_can) library
 */
#define USE_N2K_DUE_CAN 2
/**
 * \brief Use the Teensy3.x CAN Library
- [NMEA2000_teensy](https://github.com/ttlappalainen/NMEA2000_teensy) library.  
- [FlexCAN](https://github.com/ttlappalainen/FlexCAN_Library) library.  
- Remember also install [Teensyduino](https://www.pjrc.com/teensy/td_download.html) ! 
 */
#define USE_N2K_TEENSY_CAN 3 
/**
 * \brief  Use the AVR CAN Library
 - [NMEA2000_avr](https://github.com/thomasonw/NMEA2000_avr) library
 - [avr_can](https://github.com/thomasonw/avr_can) library
 */
#define USE_N2K_AVR_CAN 4   
/**
 * \brief Use the Linux Socket CAN Library
- [NMEA2000_socketCAN](https://github.com/ttlappalainen/NMEA2000_socketCAN) library.
 * NMEA2000_socketCAN has been originally forked from
 * <https://github.com/thomasonw/NMEA2000_socketCAN>
 */
#define USE_N2K_SOCKET_CAN 5 
/**
 * \brief Use the MBED CAN Library
- [NMEA2000_mbed](https://github.com/thomasonw/NMEA2000_mbed) library
other related libraries. See origin for MBED port on <https://github.com/thomasonw/NMEA2000_mbed>
 */
#define USE_N2K_MBED_CAN 6
/**
 * \brief Use the ESP32 CAN Library
 * - [NMEA2000_esp32](https://github.com/ttlappalainen/NMEA2000_esp32) library
 */
#define USE_N2K_ESP32_CAN 7   
/**
 * \brief Use the Teensy4.x CAN Library
- [NMEA2000_Teensyx](https://github.com/ttlappalainen/NMEA2000_Teensyx)  library.
- Remember also install [Teensyduino](https://www.pjrc.com/teensy/td_download.html) !
 */
#define USE_N2K_TEENSYX_CAN 8 


/*!
  \def USE_N2K_CAN
  \brief Selector for the right board
*/

// Select right CAN according to prosessor
#if !defined(USE_N2K_CAN)
#if  defined(__MBED__)					// Placing mbed 1st in tree, as the following CPUs can also be used in MBED IDE
#define USE_N2K_CAN USE_N2K_MBED_CAN
#elif defined(__SAM3X8E__)
#define USE_N2K_CAN USE_N2K_DUE_CAN
#elif defined(__ATMEGA32U4__)
#define USE_N2K_CAN USE_N2K_TEENSY_CAN
#elif defined(__MK20DX256__) || defined(__MK64FX512__) || defined (__MK66FX1M0__)
#ifndef USE_NMEA2000_TEENSYX_FOR_TEENSY_3X
  #define USE_N2K_CAN USE_N2K_TEENSY_CAN
#else
  #define USE_N2K_CAN USE_N2K_TEENSYX_CAN
#endif
#elif defined(__AVR_AT90CAN32__)||defined(__AVR_AT90CAN64__)||defined(__AVR_AT90CAN128__)|| \
      defined(__AVR_ATmega32C1__)||defined(__AVR_ATmega64C1__)||defined(__AVR_ATmega16M1__)||defined(__AVR_ATmega32M1__)|| defined(__AVR_ATmega64M1__)
#define USE_N2K_CAN USE_N2K_AVR_CAN
#elif defined(__linux__)||defined(__linux)||defined(linux)
#define USE_N2K_CAN USE_N2K_SOCKET_CAN
#elif defined(ARDUINO_ARCH_ESP32) || defined(ESP32)
#define USE_N2K_CAN USE_N2K_ESP32_CAN
#elif defined(__IMXRT1062__)
#define USE_N2K_CAN USE_N2K_TEENSYX_CAN
#else
#define USE_N2K_CAN USE_N2K_MCP_CAN
#endif
#endif


#if USE_N2K_CAN == USE_N2K_DUE_CAN
// Use Arduino Due internal CAN with due_can library
#include <NMEA2000_due.h>
#ifndef NMEA2000_ARDUINO_DUE_CAN_BUS
  #define NMEA2000_ARDUINO_DUE_CAN_BUS
#endif
tNMEA2000 &NMEA2000=*(new tNMEA2000_due(NMEA2000_ARDUINO_DUE_CAN_BUS));

#elif USE_N2K_CAN == USE_N2K_TEENSY_CAN
// Use Teensy 3.1/3.2/3.5/3.6 boards internal CAN FlexCAN library
#include <NMEA2000_teensy.h>    // https://github.com/ttlappalainen/NMEA2000_teensy
#ifndef NMEA2000_TEENSY_VER
#error Update NMEA2000_Teensy for the latest version! 
#endif
tNMEA2000 &NMEA2000=*(new tNMEA2000_teensy());

#elif USE_N2K_CAN == USE_N2K_TEENSYX_CAN
// Use Teensy 3.1/3.2/3.5/3.6/4.0/4.1 boards internal CAN
#include <NMEA2000_Teensyx.h>    // https://github.com/ttlappalainen/NMEA2000_Teensyx
tNMEA2000 &NMEA2000=*(new tNMEA2000_Teensyx(NMEA2000_TEENSYX_CAN_BUS));

#elif USE_N2K_CAN == USE_N2K_AVR_CAN
// Use Atmel AVR internal CAN controller with avr_can library
#include <avr_can.h>            // https://github.com/thomasonw/avr_can
#include <NMEA2000_avr.h>       // https://github.com/thomasonw/NMEA2000_avr
tNMEA2000 &NMEA2000=*(new tNMEA2000_avr());

#elif USE_N2K_CAN == USE_N2K_SOCKET_CAN
// Use socketCAN devices
#ifndef SOCKET_CAN_PORT
  #define SOCKET_CAN_PORT NULL
  #endif
#include <NMEA2000_SocketCAN.h>       // https://github.com/thomasonw/NMEA2000_socketCAN
tNMEA2000 &NMEA2000=*(new tNMEA2000_SocketCAN(SOCKET_CAN_PORT));
tSocketStream serStream;

#elif USE_N2K_CAN == USE_N2K_MBED_CAN
// Use MBED devices
#include <NMEA2000_mbed.h>       // https://github.com/thomasonw/NMEA2000_mbed
tNMEA2000 &NMEA2000=*(new tNMEA2000_mbed());
tmbedStream serStream;

#elif USE_N2K_CAN == USE_N2K_ESP32_CAN
#include <NMEA2000_esp32.h>       // https://github.com/ttlappalainen/NMEA2000_esp32
tNMEA2000 &NMEA2000=*(new tNMEA2000_esp32());

#else  // Use USE_N2K_MCP_CAN
// Use mcp_can library e.g. with Arduino Mega and external MCP2551 CAN bus chip
// CAN_BUS_shield libraries will be originally found on https://github.com/Seeed-Studio/CAN_BUS_Shield
// There is improved library, which branch can be found on https://github.com/peppeve/CAN_BUS_Shield
// That works also with Maple mini and 8 MHz clock. Hopefully these improvements will be applied to
// original library

#if defined(__STM32F1__) // Maple
#include <MapleIntCompatibility.h>
#endif

#include <SPI.h>
#include <mcp_can.h> // https://github.com/ttlappalainen/CAN_BUS_Shield

#if !defined(N2k_CAN_INT_PIN)
/**
 * \brief Definition of the CAN Interrupt Pin
 * This defines the number of the CAN Interrupt PIN. Default Value 0xff 
 * means not Interrupt PIN in use.
 */
#define N2k_CAN_INT_PIN 0xff   // No interrupt.
/**
 * \brief Definition of the MCP RX Buffer Size
 * Value of 1 is a small buffer size to save memory
 * 
 */
#ifndef MCP_CAN_RX_BUFFER_SIZE
#define MCP_CAN_RX_BUFFER_SIZE 1   // Just small buffer to save memory
#endif
#endif

#include <NMEA2000_mcp.h>
#if !defined(N2k_SPI_CS_PIN)
/**
 * \brief Definition of the SPI ChipSelectPin for the CAN Transceiver
 * This defines the the I/O Pin uses for the Chip Select Line to the 
 * corresponding CAN Transceiver.
 */
#define N2k_SPI_CS_PIN 53  // Pin for SPI Can Select
#endif

#if !defined(USE_MCP_CAN_CLOCK_SET)
/**
 * \brief Definition of the Clock frequency for the MCP CAN
 */
#define USE_MCP_CAN_CLOCK_SET 16
#endif

#if USE_MCP_CAN_CLOCK_SET == 8
/**
 * \brief The Clock frequency for the MCP CAN is 8MHz
 */
#define MCP_CAN_CLOCK_SET MCP_8MHz
#else
/**
 * \brief The Clock frequency for the MCP CAN is 16MHz
 */
#define MCP_CAN_CLOCK_SET MCP_16MHz
#endif

tNMEA2000 &NMEA2000=*(new tNMEA2000_mcp(N2k_SPI_CS_PIN,MCP_CAN_CLOCK_SET,N2k_CAN_INT_PIN,MCP_CAN_RX_BUFFER_SIZE));

#endif


#endif

